File: FBSimulatorControlTests/Tests/Unit/FBSimulatorConfigurationTests.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <XCTest/XCTest.h>

#import <FBSimulatorControl/FBSimulatorControl.h>

@interface FBSimulatorConfigurationTests : XCTestCase

@end

@implementation FBSimulatorConfigurationTests

- (void)testAllConfigurationsArePresent
{
  NSMutableArray<NSString *> *absentOSVersions = [NSMutableArray array];
  NSMutableArray<NSString *> *absentDeviceTypes = [NSMutableArray array];
  [FBSimulatorConfiguration allAvailableDefaultConfigurationsWithAbsentOSVersionsOut:&absentOSVersions absentDeviceTypesOut:&absentDeviceTypes];
  XCTAssertEqualObjects(absentOSVersions, @[]);
  XCTAssertEqualObjects(absentDeviceTypes, @[]);
}

- (void)testDefaultIsIphone
{
  FBSimulatorConfiguration *configuration = FBSimulatorConfiguration.defaultConfiguration;
  XCTAssertTrue([configuration.device.model containsString:@"iPhone"]);
  XCTAssertTrue([configuration.os.name containsString:@"iOS"]);
}

- (void)testiPhoneConfiguration
{
  FBSimulatorConfiguration *configuration = [[FBSimulatorConfiguration withDeviceModel:FBDeviceModeliPhone7] withOSNamed:FBOSVersionNameiOS_10_0];
  XCTAssertEqualObjects(configuration.device.model, FBDeviceModeliPhone7);
  XCTAssertEqualObjects(configuration.os.name, FBOSVersionNameiOS_10_0);
}

- (void)testiPadConfiguration
{
  FBSimulatorConfiguration *configuration = [[FBSimulatorConfiguration withDeviceModel:FBDeviceModeliPadPro] withOSNamed:FBOSVersionNameiOS_10_0];
  XCTAssertEqualObjects(configuration.device.model, FBDeviceModeliPadPro);
  XCTAssertEqualObjects(configuration.os.name, FBOSVersionNameiOS_10_0);
}

- (void)testWatchOSConfiguration
{
  FBSimulatorConfiguration *configuration = [[FBSimulatorConfiguration withDeviceModel:FBDeviceModelAppleWatchSeries2_42mm] withOSNamed:FBOSVersionNamewatchOS_3_2];
  XCTAssertEqualObjects(configuration.device.model, FBDeviceModelAppleWatchSeries2_42mm);
  XCTAssertEqualObjects(configuration.os.name, FBOSVersionNamewatchOS_3_2);
}

- (void)testTVOSConfiguration
{
  FBSimulatorConfiguration *configuration = [[FBSimulatorConfiguration withDeviceModel:FBDeviceModelAppleTV] withOSNamed:FBOSVersionNametvOS_10_0];
  XCTAssertEqualObjects(configuration.device.model, FBDeviceModelAppleTV);
  XCTAssertEqualObjects(configuration.os.name, FBOSVersionNametvOS_10_0);
}

- (void)testAdjustsOSOfIncompatableProductFamily
{
  FBSimulatorConfiguration *configuration = [[FBSimulatorConfiguration withOSNamed:FBOSVersionNametvOS_10_0] withDeviceModel:FBDeviceModeliPhone6];
  XCTAssertEqualObjects(configuration.device.model, FBDeviceModeliPhone6);
  XCTAssertTrue([configuration.os.name containsString:@"iOS"]);
}

- (void)testUsesCurrentOSIfUnknownDeviceAppears
{
  FBSimulatorConfiguration *configuration = [[FBSimulatorConfiguration withOSNamed:FBOSVersionNameiOS_10_0] withDeviceModel:@"FooPad"];
  XCTAssertEqualObjects(configuration.device.model, @"FooPad");
  XCTAssertEqualObjects(configuration.os.name, FBOSVersionNameiOS_10_0);
}

- (void)testUsesCurrentDeviceIfUnknownOSAppears
{
  FBSimulatorConfiguration *configuration = [[FBSimulatorConfiguration withDeviceModel:FBDeviceModeliPhone7] withOSNamed:@"FooOS"];
  XCTAssertEqualObjects(configuration.device.model, FBDeviceModeliPhone7);
  XCTAssertEqualObjects(configuration.os.name, @"FooOS");
}

@end

File: FBSimulatorControlTests/Tests/Unit/FBSimulatorConfigurationTests.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import FBControlCore
@testable import FBSimulatorControl
import XCTest

final class FBSimulatorConfigurationTests: XCTestCase {

  override func setUpWithError() throws {
    try super.setUpWithError()
    // Every test here seeds from `defaultConfiguration()`, which resolves the newest runtime
    // *installed on this host* that supports its device family. A host with no simulator runtimes
    // cannot satisfy that, so these need one rather than being pure unit tests. Skip explicitly
    // instead of failing, so the suite distinguishes "not exercised here" from "broken".
    if (try? FBSimulatorConfiguration.defaultConfiguration()) == nil {
      throw XCTSkip("Requires at least one installed simulator runtime")
    }
  }

  func testDefaultIsIphone() throws {
    let configuration = try FBSimulatorConfiguration.defaultConfiguration()
    XCTAssertTrue(configuration.device.model.rawValue.contains("iPhone"))
    XCTAssertTrue(configuration.os.name.rawValue.contains("iOS"))
  }

  func testiPhoneConfiguration() throws {
    let configuration = try FBSimulatorConfiguration.defaultConfiguration().withDeviceModel(.modeliPhone7).withOSNamed(.nameiOS_10_0)
    XCTAssertEqual(configuration.device.model, .modeliPhone7)
    XCTAssertEqual(configuration.os.name, .nameiOS_10_0)
  }

  func testiPadConfiguration() throws {
    let configuration = try FBSimulatorConfiguration.defaultConfiguration().withDeviceModel(.modeliPadPro).withOSNamed(.nameiOS_10_0)
    XCTAssertEqual(configuration.device.model, .modeliPadPro)
    XCTAssertEqual(configuration.os.name, .nameiOS_10_0)
  }

  func testWatchOSConfiguration() throws {
    let watchModel = FBDeviceModel(rawValue: "Apple Watch Series 2 - 42mm")
    let watchOS = FBOSVersionName(rawValue: "watchOS 3.2")
    let configuration = try FBSimulatorConfiguration.defaultConfiguration().withDeviceModel(watchModel).withOSNamed(watchOS)
    XCTAssertEqual(configuration.device.model, watchModel)
    XCTAssertEqual(configuration.os.name, watchOS)
  }

  func testTVOSConfiguration() throws {
    let tvModel = FBDeviceModel(rawValue: "Apple TV")
    let tvOS = FBOSVersionName(rawValue: "tvOS 10.0")
    let configuration = try FBSimulatorConfiguration.defaultConfiguration().withDeviceModel(tvModel).withOSNamed(tvOS)
    XCTAssertEqual(configuration.device.model, tvModel)
    XCTAssertEqual(configuration.os.name, tvOS)
  }

  func testAdjustsOSOfIncompatableProductFamily() throws {
    let tvOS = FBOSVersionName(rawValue: "tvOS 10.0")
    let configuration = try FBSimulatorConfiguration.defaultConfiguration().withOSNamed(tvOS).withDeviceModel(.modeliPhone6)
    XCTAssertEqual(configuration.device.model, .modeliPhone6)
    XCTAssertTrue(configuration.os.name.rawValue.contains("iOS"))
  }

  func testUsesCurrentOSIfUnknownDeviceAppears() throws {
    let configuration = try FBSimulatorConfiguration.defaultConfiguration().withOSNamed(.nameiOS_10_0).withDeviceModel(FBDeviceModel(rawValue: "FooPad"))
    XCTAssertEqual(configuration.device.model, FBDeviceModel(rawValue: "FooPad"))
    XCTAssertEqual(configuration.os.name, .nameiOS_10_0)
  }

  func testUsesCurrentDeviceIfUnknownOSAppears() throws {
    let configuration = try FBSimulatorConfiguration.defaultConfiguration().withDeviceModel(.modeliPhone7).withOSNamed(FBOSVersionName(rawValue: "FooOS"))
    XCTAssertEqual(configuration.device.model, .modeliPhone7)
    XCTAssertEqual(configuration.os.name, FBOSVersionName(rawValue: "FooOS"))
  }
}
